### ISSUE TITLE
A proposal for chain initialization, main chain block inclusion, and deposit processing

### ISSUE THREAD

**Issue by vbuterin** [ethereum/eth2.0-specs#71]
### On-chain contract

We create a deposit contract on the blockchain, with roughly the following code:

```python
HashChainValue: event({prev_tip: bytes32, data: bytes[2048], value: wei_value, total_deposit_count: int128})
ChainStart: event({hash_chain_tip: bytes32, time: timestamp})

hash_chain_tip: public(bytes32)
total_deposit_count: int128

@payable
@public
def deposit(data: bytes[2048]):
    log.HashChainValue(self.hash_chain_tip, data, msg.value, self.total_deposit_count)
    self.total_deposit_count += 1
    self.hash_chain_tip = sha3(concat(self.hash_chain_tip, data, as_bytes32(msg.value), as_bytes32(self.total_deposit_count)))
    if self.total_deposit_count == 16384:
        log.ChainStart(self.hash_chain_tip, block.timestamp)
```

When a user wishes to move their ETH from the 1.0 chain to the 2.0 chain, they should call the `deposit` function, sending along 32 ETH and providing as `data` a SimpleSerialize'd object with the following arguments (in order):

* `pubkey`: int256
* `proof_of_possession`: [int256]
* `withdrawal_shard`: int64
* `withdrawal_address`: bytes20
* `randao_commitment`: hash32

If they wish to deposit more than 32 ETH, they would need to make multiple calls.

[Governance note: when publishing the final version of this contract, it may be desirable to issue some kind of formal EIP that "enshrines" its privileged status, ąnd encourage CarbonVotes and other polls to vote on it, which if successful wouuld make clear to the community that this contract is "part of the protocol" and so the community *is* responsible for providing ongoing protocol improvements that eventually unlock any ETH sent into this contract]

### Chain initialization

When a `ChainStart` log is published, this initializes the chain, setting the following parameters:

* `POW_CHAIN_HASH_ROOT` (new parameter) = `hash_chain_tip`
* `GENESIS_TIME` = `time`
* `PROCESSED_HASH_ROOT` (new parameter) = `hash_chain_tip`

It runs `on_startup` with `initial_validator_entries` equal to the list of data records published as HashChainValue logs so far, in the order in which they were published (oldest to newest).

### Chain updating

Define a validator's "view" as being the value obtained by calling `DEPOSIT_CONTRACT_ADDRESS.get_hash_chain_tip()` from the post-state of the block 512 blocks behind the current head of the PoW chain. Define a "valid view" (defined subjectively from the PoW of a validator) as a value which is a descendant of `POW_CHAIN_HASH_ROOT` and cannot be obtained by calling `DEPOSIT_CONTRACT_ADDRESS.get_hash_chain_tip()` from the post-state of a block that is part of the canonical PoW chain _at least_ 512 blocks behind the head. Note that any valid view should be either equal to or an ancestor of the validator's view.

Blocks will have a new data field, `hash_chain_tip_vote: hash32`, which proposers are expected to fill with the following algorithm:

* Let slot B be the last slot during which the `POW_CHAIN_HASH_ROOT` changed.
* If all blocks since slot B contained a `hash_chain_tip_vote` that was either equal to the `POW_CHAIN_HASH_ROOT` or was an invalid view, vote the validator's view.
* If there was at least one valid view published as a `hash_chain_tip_vote` since slot B, copy the first valid view.

Note that assuming >= 50% honest, this algorithm will converge to all honest proposers voting the same value, which is a descendant of the `POW_CHAIN_HASH_ROOT`. If the same value is voted for in >= 683 of the last 1024 blocks, set the `POW_CHAIN_HASH_ROOT` to this value.

**Note that this is a vote, not a consensus rule**; blocks with incorrect votes should not be rejected.

### Deposit processing

Add a new type of `SpecialObject`, which consists of the entire hash-linked-list of HashChainValue logs since the previous `PROCESSED_HASH_ROOT` up to the `POW_CHAIN_HASH_ROOT`. If a valid such hash-linked-list is submitted, then we run `add_validator` with the given values for each record, and set `PROCESSED_HASH_ROOT = POW_CHAIN_HASH_ROOT`.

This does mean that deposit processing is not "automatic", in that deposits are not automatically read from the PoW chain; only a hash is automatically read, and the rest of the data must be manually imported by some block proposer. This is by design, to limit the amount of in-consensus communication between the PoW chain and the beacon chain required to a single hash value.

**Comment by jannikluhn** [comment]
If I understand correctly, this allows the first proposer submitting a new valid vote to delay deposit processing by 511 additional blocks, as they can vote on the child of `POW_CHAIN_HASH_ROOT` instead of their actual view and all following proposers would do the same.

I'm wondering if we could fix this by counting votes for a view also towards all ancestors of that view. Then proposers wouldn't have to copy the first vote but could always vote for their view. `POW_CHAIN_HASH_ROOT` would be updated continuously (but still delayed by 512 blocks compared to the head) and there wouldn't be a special first vote. Sounds too simple, what's the catch?

**Comment by vbuterin** [comment]
> If I understand correctly, this allows the first proposer submitting a new valid vote to delay deposit processing by 511 additional blocks, as they can vote on the child of POW_CHAIN_HASH_ROOT instead of their actual view and all following proposers would do the same.

Correct. But all that means is that if < 50% of proposers are dishonest, deposit processing will on average be delayed by < 512 additional blocks. So it's not that bad. As soon as you get lucky with one honest proposer, the chain can catch up fully.

> I'm wondering if we could fix this by counting votes for a view also towards all ancestors of that view

The problem is that this would require the beacon chain to be aware of the concept of ancestors within that mechanism, which would add considerable complexity.

**Comment by JustinDrake** [comment]
Minor suggestions for the deposit contract:

1. Update the size of `pubkey` and `proof_of_possession` for the BLS12-381 curve.
2. Do not cap the number of bytes in `data` to `300` to be more update-friendly.
3. Consider removing `assert msg.value == as_wei_value(32, "ether")` as this is not friendly to deposit top ups, deposit buffers, or other future changes related to the size of the deposit.
4. Consider changing the constant `10000` to a power of two for consistency (also, name it and add it to the list of constants in the spec).
5. Consider hashing `total_deposit_count` into `hash_chain_tip` so that the `total_deposit_count` can be readily used by light clients. Bonus: this removes the need for the `ChainStart` event and the `self.total_deposit_count == 10000` check.

**Comment by djrtwo** [comment]
Should we consider adding a small proposer reward for adding a well-formatted `SpecialRecord` that triggers `add_validator`?

**Comment by vbuterin** [comment]
@JustinDrake fixed all except (1) [probably better to run through and update to BLS12-381 all at once at some point]
@djrtwo sounds good to me!

**Comment by vbuterin** [comment]
Here is an alternative that allows for the same structure to be used for accepting deposits from the PoW chain *and* from shards.

The code changes to:

```
HashChainValue: event({prev_tip: bytes32, data: bytes[2048], value: wei_value, total_deposit_count: int128})
ChainStart: event({hash_chain_tip: bytes32, time: timestamp})

receipt_tree: bytes32[int128]
total_deposit_count: int128

@payable
@public
def deposit(data: bytes[2048]):
    log.HashChainValue(self.hash_chain_tip, data, msg.value, self.total_deposit_count)
    index:int128 = self.total_deposit_count + 2**32
    self.receipt_tree[index] = sha3(concat(data, as_bytes32(msg.value), as_bytes32(block.timestamp))
    for i in range(32):
        index //= 2
        self.receipt_tree[index] = sha3(concat(self.receipt_tree[index * 2], self.receipt_tree[index * 2 + 1]))
    self.total_deposit_count += 1
    if self.total_deposit_count == 16384:
        log.ChainStart(self.receipt_tree[1], block.timestamp)

@public
@constant
def get_receipt_root() -> bytes32:
    return self.receipt_tree[1]
```

Now, any receipt can prove its membership separately from a Merkle tree. We also make a few changes:

* A receipt can be used to deposit if the beacon chain time is less than `max(receipt.timestamp, chain.genesis_time) + DEPOSIT_WINDOW` (eg. 2^25 seconds ~= 1 year).
* Withdrawn validator slots stay in the state for 1 year
* The proof of possession is now a signature of not just the hash of the pubkey, but of the pubkey+data. This prevents attackers from griefing by submitting a deposit with the same pubkey but a different randao hash root or withdrawal address.
* A rule is added that you cannot deposit a validator that has the same pubkey as any validator currently in the state. This prevents double-spending receipts.

Notice that this mechanism can also later be used unchanged to import validators from shards.

**Comment by djrtwo** [comment]
Log `HashChainValue` is still using `hash_chain_tip`. Should this be `self.receipt_tree[1]` in the 0th arg?

**Comment by djrtwo** [comment]
Why the one year? Is this to compensate for the fact that withdrawals can now happen fast in the beacon chain (if only a short queue) to prevent quickly reregistering of public key after withdrawal?

**Comment by vbuterin** [comment]
> Log HashChainValue is still using self.hash_chain_tip. Should this be self.receipt_tree[1] in the 0th arg?

Ah, yes it should.

> Why the one year? Is this to compensate for the fact that withdrawals can now happen fast in the beacon chain (if only a short queue) to prevent quickly reregistering of public key after withdrawal?

It could be any length. But if the length is short, then it creates risk of loss of funds if the beacon chain breaks for an extended duration.

**Comment by djrtwo** [comment]
closed via #107
